### PR TITLE
feat(kustomize): override namespace (#6067)

### LINKED ISSUE #6067: Support Dynamic Namespaces via Kustomize
# Summary

Add support for `kustomize edit set namespace <name>` in `Application` CRD.

# Motivation

`Application.spec.destination.namespace` is often insufficient for deploying to dynamic namespaces. Kubectl can set objects' `metadata.namespace` (only if unset,) but it does not set `ClusterRoleBinding.subjects[*].namespace`, `APIService.spec.service.namespace`, namespace in CustomResource objects, and more. Kustomize handles changing the namespace in all the right places, even custom places via transformerConfigs, all by setting `namespace: <name>` in the kustomization.yaml.

This would better support things like namespaced PR based ephemeral environments, client-per-namespace environments, and those of us who like to explicitly set namespace in our configs rather than use the one in the kubeconfig.

# Proposal

ArgoCD already implements `kustomize edit set ...` commands, to change `commonAnnotations`, `commonLabels`, `images`, `namePrefix`, and `nameSuffix`, all configured via `Application.spec.source.kustomize.<config>`. I propose we implement `Application.spec.source.kustomize.namespace` which runs `kustomize edit set namespace <name>`.

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
Fixes #6067

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
@AndrewChubatiuk this is awesome! Care to take a crack at UI/CLI support? If not, we can plan to write a follow-up issue for those.

File: docs/user-guide/commands/argocd_app_set.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ argocd app set APPNAME [flags]
       --kustomize-force-common-annotation          Force common annotations in Kustomize
       --kustomize-force-common-label               Force common labels in Kustomize
       --kustomize-image stringArray                Kustomize images (e.g. --kustomize-image node:8.15.0 --kustomize-image mysql=mariadb,alpine@sha256:24a0c4b4a4c0eb97a1aabb8e29f18e917d05abfe1b7a7c07857230879ce7d3d)
+      --kustomize-namespace string                 Kustomize namespace
```
Comment (crenshaw-dev) [inline docs/user-guide/commands/argocd_app_set.md:41]: Do you have time to add the corresponding `argocd app unset` parameter?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: pkg/apis/application/v1alpha1/types.go [hunk 2]
```diff
@@ -381,13 +381,16 @@ type ApplicationSourceKustomize struct {
 	ForceCommonLabels bool `json:"forceCommonLabels,omitempty" protobuf:"bytes,7,opt,name=forceCommonLabels"`
 	// ForceCommonAnnotations specifies whether to force applying common annotations to resources for Kustomize apps
 	ForceCommonAnnotations bool `json:"forceCommonAnnotations,omitempty" protobuf:"bytes,8,opt,name=forceCommonAnnotations"`
+	// Namespace overrides Kustomize namespace
```
Comment (crenshaw-dev) [inline pkg/apis/application/v1alpha1/types.go:384]:
```suggestion
	// Namespace sets the namespace that Kustomize adds to all resources
```

I think this is slightly more clear?

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
Looks amazing. One last nitpick.

I'm also going to look at the original issue to see if there are any common use cases we should document. But otherwise lgtm.

**Comment by crenshaw-dev** [comment]
What do you think of adding a new section at the bottom of kustomize.md?

> ### Setting the manifests' namespace
> 
> The `spec.destination.namespace` field only adds a namespace when it's missing from the manifests generated by Kustomize. It also uses `kubectl` to set the namespace, which sometimes misses namespace fields in certain resources (for example, custom resources). In these cases, you might get an error like this: `ClusterRoleBinding.rbac.authorization.k8s.io "example" is invalid: subjects[0].namespace: Required value`.
> 
> Using Kustomize directly to set the missing namespaces can resolve this problem. Setting `spec.source.kustomize.namespace` instructs Kustomize to set namespace fields to the given value. 
> 
> If `spec.destination.namespace` and `spec.source.kustomize.namespace` are both set, Argo CD will defer to the latter, the namespace value set by Kustomize.

**Review by ishitasequeira — APPROVED** [review]
LGTM!!

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]
@AndrewChubatiuk if I set just `kustomize.namespace` in the `Manifest` tab of the Application Details pane, the entire `kustomize` block gets deleted. I see you updated `IsZero`, so I'm not sure why this is happening.  :-(

**Comment by crenshaw-dev** [comment]
@AndrewChubatiuk Intuit is going to give me time to make sure this gets into 2.7. I'll circle back when it gets added to my sprint, sometime in the next ~1mo. Thanks for your patience!

**Comment by crenshaw-dev** [comment]
@AndrewChubatiuk I finally have time. :-) Can you resolve conflicts? If you don't have time, I'll see if GitHub will let me push to your branch.

**Comment by crenshaw-dev** [comment]
I think I managed to rebase.

File: Makefile [hunk 3]
```diff
@@ -219,7 +219,7 @@ clidocsgen: ensure-gopath
 
 
 .PHONY: codegen-local
-codegen-local: ensure-gopath mod-vendor-local notification-docs notification-catalog gogen protogen clientgen openapigen clidocsgen manifests-local
+codegen-local: ensure-gopath mod-vendor-local gogen protogen clientgen openapigen clidocsgen manifests-local
```
Comment (crenshaw-dev) [inline Makefile:222]:
```suggestion
codegen-local: ensure-gopath mod-vendor-local notification-docs notification-cataloggogen protogen clientgen openapigen clidocsgen manifests-local
```

File: Makefile [hunk 4]
```diff
@@ -219,7 +219,7 @@ clidocsgen: ensure-gopath
 
 
 .PHONY: codegen-local
-codegen-local: ensure-gopath mod-vendor-local notification-docs notification-catalog gogen protogen clientgen openapigen clidocsgen manifests-local
+codegen-local: ensure-gopath mod-vendor-local notification-docs notification-cataloggogen protogen clientgen openapigen clidocsgen manifests-local
```
Comment (crenshaw-dev) [inline Makefile:222]:
```suggestion
codegen-local: ensure-gopath mod-vendor-local notification-docs notification-catalog gogen protogen clientgen openapigen clidocsgen manifests-local
```

**Review by crenshaw-dev — APPROVED** [review]
Thanks for your work and for your patience, @AndrewChubatiuk!

**Comment by snarlysodboxer** [comment]
@AndrewChubatiuk Awesome work on this, thank you!! This allows us to remove a ton of complexity in our configs. :)

**Pull request merged**